pep-0612.rst

```diff
@@ -0,0 +1,258 @@
+
+Rejected Alternatives
+---------------------
+
+Using List Variadics and Map Variadics
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We considered just trying to make something like this with a callback protocol which was parameterized on a list-type variadic, and a map-type variadic like so:
+
+.. code-block::
+
+   Treturn = typing.TypeVar(“Treturn”)
+   Tpositionals = ....
+   Tkeywords = ...
+   class BetterCallable(typing.Protocol[Tpositionals, Tkeywords, Treturn]):
+     def __call__(*args: Tpositionals, **kwargs: Tkeywords) -> Treturn: ...
+
+However there are some problems with trying to come up with a consistent solution for those type variables for a given callable. This problem comes up with even the simplest of callables:
+
+.. code-block::
+
+   def simple(x: int) -> None: ...
+   simple <: BetterCallable[[int], [], None]
+   simple <: BetterCallable[[], {“x”: int}, None]
+   BetterCallable[[int], [], None] </: BetterCallable[[], {“x”: int}, None]
+
+Any time where a type can implement a protocol in more than one way that aren’t mutually compatible, we can run into situations where we lose information. If we were to make a decorator using this protocol, we have to pick one calling convention to prefer.
+
+.. code-block::
+
+   def decorator(
+     f: BetterCallable[[Ts], [Tmap], int],
+   ) -> BetterCallable[[Ts], [Tmap], str]:
+       def decorated(*args: Ts, **kwargs: Tmap) -> str:
+          x = f(*args, **kwargs) 
+          return int_to_str(x)
+       return decorated
+   @decorator
+   def foo(x: int) -> int:
+       return x
+   reveal_type(foo) # Option A: BetterCallable[[int], {}, str]
+                    # Option B: BetterCallable[[], {x: int}, str]
+   foo(7)   # fails under option B
+   foo(x=7) # fails under option A
+
+The core problem here is that, by default, parameters in Python can either be passed in positionally or as a keyword parameter. This means we really have three categories (positional-only, positional-or-keyword, keyword-only) we’re trying to jam into two categories. This is the same problem that we briefly mentioned when discussing ``.args`` and ``.kwargs``. Fundamentally, in order to capture two categories when there are some things that can be in either category, we need a higher level primitive (\ ``ParameterSpecification``\ ) to capture all three, and then split them out afterward.
+
+Mutations on ParameterSpecifications
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There are still a class of decorators still not supported with these features: those that mutate (add/remove/change) the parameters of the given function. Defining operators that do these mutations becomes very complicated very quickly, as you have to deal with name collision issues much more prominently. We will defer that work until there is significant demand, and then we would be open to revisiting it.
```

Comment:
I would say that this is needed, there is even a whole family of decorators that do just that. It seems like `ParameterSpecification` does not leave an open space for this feature in the future. And can even be conflicting with it.

Real world examples of decorators that do change both input and output types: https://github.com/dry-python/returns/blob/master/returns/contrib/mypy/decorator_plugin.py

Reply:
Would you mind explaining this example in a bit more detail?  To me it appears that those decorators are propagating the entire set of parameters over, without any mutation. 

By mutation of parameters, I was referring to something like this: (removal)
```
def decorator(f):
    def inner(*args, **kwargs):
         return f(extra=1, *args, **kwargs)
    return inner
```
or this (addition)
```
def decorator(f):
    def inner(extra, *args, **kwargs):
         return f(*args, **kwargs)
    return inner
```

Reply:
In this particular case we often change both input and return types: `T` -> `Wrapper[T]`, where wrapper can be `List`, `Optional`, `Maybe`, `IO`, etc.

Reply:
Are you referring to a decorator which turns a `Callable[[int, str], bool]` into a `Callable[[Wrapper[int], Wrapper[str]], Wrapper[bool]]`?  
Would you mind giving a me a snippet or pointer to an instance of this so I can better understand what you're describing here?

Reply:
Sure: 

- https://returns.readthedocs.io/en/latest/pages/maybe.html#maybe-decorator
- https://returns.readthedocs.io/en/latest/pages/io.html#impure
- https://returns.readthedocs.io/en/latest/pages/result.html#safe

Reply:
All three of those appear to be only modifying the return types of the decorated functions.
This means, that with ParameterSpecifications, you should be able to type them as 
```
TParams = typing.ParameterSpecification("TParams")
def maybe(
    function: Callable[TParams, Optional[_ValueType]],
) -> Callable[TParams, Maybe[_ValueType]]: ...
```

```
def impure(
    function: Callable[TParams, _NewValueType],
) -> Callable[TParams, IO[_NewValueType]]:....
```

```
def safe(
    function: Callable[TParams, _ValueType],
) -> Callable[TParams, Result[_ValueType, Exception]]: ...
```

with overloads for coroutines like
```
def maybe(
    function: Callable[
        TParams,
        Coroutine[_FirstType, _SecondType, Optional[_ValueType]],
    ],
) -> Callable[
    TParams,
    Coroutine[_FirstType, _SecondType, Maybe[_ValueType]],
]:
```
and so on for the other ones.

Will that work for your use cases?

Reply:
Fair enough! This would work for my cases. 

Thanks a lot for taking your time and explaining this. 👍

Reply:
No problem!  Thanks for taking a look at the PEP.  As @Rosuav said, if you have any other questions/comments/concerns about this, we should discuss them on typing-sig@python.org (https://mail.python.org/archives/list/typing-sig@python.org/thread/UDHSH4EVVIDKNLRX2YGCIUCBGZ5ALRKC/)